### PR TITLE
monomorphisation ain't free, part 2: reduce # of `RelationWrapper` template instances in synth

### PR DESCRIPTION
**Pull request by ohamel-softwaresecure**
Change: Remove non-type template params from `RelationWrapper` to eliminate most equiv instances generated when compiling synth'd code.

Motivation: `std::unique_ptr` takes a whooping ~5ms to instantiate on clang-10. This adds up w/ 1000s of tiny parameterised types (e.g. synth'd code). (1k relations, 5 seconds *just* to instance templates template. doesn't include CG & opt). Using inner classes or type defs also inhibits memoisation, even if the typedef or class only depends on a subset of the parent template's type parameters.

Future Work: The API exposed by `RelationWrapper` is primarily for the benefit of external C++ and it isn't a hot code path. It would be possible to completely remove the templating for `RelationWrapper` if we added a dynamic dispatch to `RelType`.

### CODE REVIEW

**Review by mmcgr — COMMENTED** [review]
Only a minor quibble from me, but I would like @b-scholz 's approval for interface changes.

File: src/include/souffle/SouffleInterface.h [hunk 1]
```diff
@@ -41,6 +41,10 @@ class tuple;
  * Object-oriented wrapper class for Souffle's templatized relations.
  */
 class Relation {
+public:
+    // you have bigger fish to fry if you have > 255 attributes
+    using arity_type = uint8_t;
```
Comment (b-scholz) [inline src/include/souffle/SouffleInterface.h:46]: Is there a reason to limit it uint8_t? Could we have uint32_t instead?

File: src/include/souffle/SouffleInterface.h [hunk 2]
```diff
@@ -695,21 +727,21 @@ class SouffleProgram {
...
      * @param name the name of the relation (std::string)
-     * @param rel a pointer to the relation (std::string)
+     * @param rel a reference to the relation
      * @param isInput a bool argument, true if the relation is a input relation, else false (bool)
      * @param isOnput a bool argument, true if the relation is a ouput relation, else false (bool)
      */
-    void addRelation(const std::string& name, Relation* rel, bool isInput, bool isOutput) {
-        relationMap[name] = rel;
-        allRelations.push_back(rel);
+    void addRelation(const std::string& name, Relation& rel, bool isInput, bool isOutput) {
```
Comment (b-scholz) [inline src/include/souffle/SouffleInterface.h:734]: Is this change necessary? It will cause an interface change; what is the rationale?

Comment (ohamel-softwaresecure) [reply src/include/souffle/SouffleInterface.h:734]: Rationale: Safety - You can't pass in a null reference.
I'll add a deprecated overload for back-compat.

File: src/include/souffle/SouffleInterface.h (hunk 1 above)
Comment (ohamel-softwaresecure) [reply src/include/souffle/SouffleInterface.h:46]: We could, but is there a reason to? It was meant more as a cheap sancheck/type-assertion that we haven't considered relations with absurdly massive tuple.

Comment (b-scholz) [reply src/include/souffle/SouffleInterface.h:46]: I can think of examples (transposes of relations) where more than 256 attributes may be possible.
Hence, I don't want to restrict arities to 8bit, artificially.

File: src/include/souffle/SouffleInterface.h (hunk 2 above)
Comment (b-scholz) [reply src/include/souffle/SouffleInterface.h:734]: Great.

File: src/include/souffle/SouffleInterface.h (hunk 1 above)
Comment (ohamel-softwaresecure) [reply src/include/souffle/SouffleInterface.h:46]: I have no idea how they live with the compile times for the data structures.

**Comment by mmcgr** [comment]
The style problem seems buggy. PR #1717 will, indirectly, fix this.

**Review by b-scholz — APPROVED** [review]
